zarr/core.py

```diff
@@ -1714,6 +1740,8 @@ def _chunk_setitem_nosync(self, chunk_coords, chunk_selection, value, fields=Non
             else:
                 chunk[chunk_selection] = value
 
+        if self._chunk_cache is not None:
+            self._chunk_cache[ckey] = np.copy(chunk)
```

Comment:
To solve the test_object_arrays_vlen_array failure instead of the following two lines:
```python
        if self._chunk_cache is not None:
            self._chunk_cache[ckey] = np.copy(chunk) 
```

I could do this(have tried and test passes):
```python
        import numcodecs  # will put this import up of course
        if self._chunk_cache is not None:
            if self._filters:
                for f in self._filters:
                    if type(f) is numcodecs.VLenArray:
                        chunk = f.decode(f.encode(chunk))
            self._chunk_cache[ckey] = np.copy(chunk) 
```
This does look quite messy though.

We could do the above or we could expect the user to provide correct dtypes of arrays when using write cache and VLenArray codec, and skip the test in the build. I'm not sure how else to resolve this test failure.